### PR TITLE
Notebook support

### PR DESCRIPTION
**Pull request by arikrundquist**
This pr adds basic support for mypy diagnostic messages in jupyter notebooks. Since dmypy does not natively support type checking notebooks, we resort to concatenating the python cells in a notebook to a single string, and then check that string using the mypy `-c <string>` command line option.

This implements feature request #94.

image

### CODE REVIEW

**Comment by matangover** [comment]
Awesome!! Could you please revert unnecessary whitespace changes to make this easier to review?

**Comment by matangover** [comment]
(You changed all the indentation for some reason)

**Comment by arikrundquist** [comment]
Good catch, I reverted all the other auto formatting changes but I always have my whitespace diff turned off. Should be fixed now.

**Comment by matangover** [comment]
Great. Thanks

File: src/extension.ts [hunk 1]
```diff
@@ -407,6 +420,148 @@ async function getDmypyExecutable(folder: vscode.Uri, warnIfFailed: boolean, cur
 	return dmypyExecutable;
 }
 
+async function checkAllNotebooks() {
+	for (const notebook of vscode.workspace.notebookDocuments) {
+		await checkNotebook(notebook)
+	}
```
Comment (Quasar6X) [inline src/extension.ts:426]: This should be marginally faster given that `Promise.all` can enque all the `Promises` at once on the eventloop.

```suggestion
	await Promise.all(vscode.workspace.notebookDocuments.map(notebook => checkNotebook(notebook)));
```

**Comment by matangover** [comment]
Sorry for delay in reviewing, hope to do it in the next few days

**Comment by arikrundquist** [comment]
just wanted to follow up and check in on the status of this pr

**Comment by matangover** [comment]
Sorry again, I was busy. I just finished something big so should have more time

**Comment by matangover** [comment]
I really like what you did here. Thanks for the PR. I am refactoring it slightly and adding a flag `mypy.checkNotebooks` which, at least for now, will be disabled by default. I will do some testing and then release it.

**Comment by matangover** [comment]
Also `mypy.configFile` and `mypy.extraArguments`

**Comment by matangover** [comment]
I had some more time to finish this up. I wanted the extension to get the selected notebook interpreter and pass that to mypy using --python-executable (as I do for normal python files). However turns out this is a real rabbit hole. The Jupyter extension has an API to get the selected interpreter path, but the API is restricted only to approved developers (what??). So it doesn't seem possible to get the selected interpreter path for a notebook. Unfortunately this means that notebook checking will not always be correct (e.g. missing packages even though they are installed). Perhaps a workaround could be found at some point. For now the extension will assume the notebook is using the same interpreter as the workspace.

**Comment by arikrundquist** [comment]
Oh that's actually so bizarre that the api is restricted. It's unfortunate that means the python interpreter could be wrong, but hopefully for most cases it will still be helpful (I personally always configure both vscode and jupyter to use the same interpreter anyway, but I don't really use notebooks enough to have a good sense of if/when that might not make sense to do).

Thanks for looking into this, I hadn't even considered that the workspace and notebook interpreters could be different, so even though we're kind of stuck with where it's at I'm glad someone at least thought about that potential issue.

**Comment by matangover** [comment]
I also set the workspace interpreter and notebook interpreter to almost always be the same. So yeah this should be ok :)

**Pull request merged**